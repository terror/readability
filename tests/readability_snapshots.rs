use {
  readability::{Readability, ReadabilityOptions},
  std::{
    fs,
    path::{Path, PathBuf},
  },
};

fn load_fixture(path: &Path) -> anyhow::Result<String> {
  Ok(fs::read_to_string(path.join("source.html"))?)
}

fn run_fixture(name: &str, path: PathBuf) -> anyhow::Result<()> {
  let html = load_fixture(&path)?;

  let mut parser =
    Readability::new(&html, None, ReadabilityOptions::default())?;

  let article = parser.parse()?;

  insta::with_settings!({ snapshot_suffix => name }, {
    insta::assert_json_snapshot!("article", &article);
  });

  Ok(())
}

fn fixture_dirs() -> Vec<(String, PathBuf)> {
  let mut fixtures = Vec::new();

  let root = PathBuf::from("tests/fixtures");

  if root.exists() {
    for entry in fs::read_dir(&root).expect("fixtures directory") {
      let entry = entry.expect("fixture entry");

      if entry.file_type().map(|ty| ty.is_dir()).unwrap_or(false) {
        let path = entry.path();

        if path.join("source.html").exists() {
          let name = entry.file_name().to_string_lossy().to_string();
          fixtures.push((name, path));
        }
      }
    }
  }

  if let Ok(env_root) = std::env::var("READABILITY_TEST_ROOT") {
    let env_path = PathBuf::from(env_root);

    if env_path.exists() {
      for entry in walkdir::WalkDir::new(env_path)
        .max_depth(2)
        .into_iter()
        .filter_map(Result::ok)
        .filter(|e| e.file_type().is_dir())
      {
        let dir = entry.path().to_path_buf();

        if dir.join("source.html").exists() {
          let name = dir
            .file_name()
            .unwrap_or_default()
            .to_string_lossy()
            .to_string();

          fixtures.push((name, dir));
        }
      }
    }
  }

  fixtures
}

#[test]
fn readability_snapshots() -> anyhow::Result<()> {
  let fixtures = fixture_dirs();

  if fixtures.is_empty() {
    return Ok(());
  }

  for (name, path) in fixtures {
    run_fixture(&name, path)?;
  }

  Ok(())
}
