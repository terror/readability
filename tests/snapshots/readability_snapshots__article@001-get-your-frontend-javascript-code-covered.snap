---
source: tests/readability_snapshots.rs
expression: "&article"
---
{
  "title": "Get your Frontend JavaScript Code Covered | Code | Nicolas Perriault",
  "byline": "Nicolas Perriault",
  "dir": null,
  "lang": "en",
  "content": "<div id=\"readability-content\"><div id=\"readability-page-1\" class=\"page\"><section>\n                        <p><strong>So finally you're <a href=\"/code/2013/testing-frontend-javascript-code-using-mocha-chai-and-sinon/\">testing your frontend JavaScript code</a>? Great! The more you\nwrite tests, the more confident you are with your code… but how much precisely?\nThat's where <a href=\"http://en.wikipedia.org/wiki/Code_coverage\">code coverage</a> might\nhelp.</strong>\n                        </p>\n                        <p>The idea behind code coverage is to record which parts of your code (functions,\n                            statements, conditionals and so on) have been executed by your test suite,\n                            to compute metrics out of these data and usually to provide tools for navigating\n                            and inspecting them.</p>\n                        <p>Not a lot of frontend developers I know actually test their frontend code,\n                            and I can barely imagine how many of them have ever setup code coverage…\n                            Mostly because there are not many frontend-oriented tools in this area\n                            I guess.</p>\n                        <p>Actually I've only found one which provides an adapter for <a href=\"http://visionmedia.github.io/mocha/\">Mocha</a> and\n                            actually works…</p>\n                        <blockquote class=\"twitter-tweet tw-align-center\">\n                            <p>Drinking game for web devs:\n                                <br>(1) Think of a noun\n                                <br>(2) Google \"&lt;noun&gt;.js\"\n                                <br>(3) If a library with that name exists - drink</p>— Shay Friedman (@ironshay)\n                            <a href=\"https://twitter.com/ironshay/statuses/370525864523743232\">August 22, 2013</a>\n                        </blockquote>\n                        <p><strong><a href=\"http://blanketjs.org/\">Blanket.js</a></strong> is an <em>easy to install, easy to configure,\nand easy to use JavaScript code coverage library that works both in-browser and\nwith nodejs.</em>\n                        </p>\n                        <p>Its use is dead easy, adding Blanket support to your Mocha test suite\n                            is just matter of adding this simple line to your HTML test file:</p>\n<pre><code>&lt;script src=\"vendor/blanket.js\"\n        data-cover-adapter=\"vendor/mocha-blanket.js\"&gt;&lt;/script&gt;\n</code></pre>\n\n                        <p>Source files: <a href=\"https://raw.github.com/alex-seville/blanket/master/dist/qunit/blanket.min.js\">blanket.js</a>,\n                            <a href=\"https://raw.github.com/alex-seville/blanket/master/src/adapters/mocha-blanket.js\">mocha-blanket.js</a>\n                        </p>\n                        <p>As an example, let's reuse the silly <code>Cow</code> example we used\n                            <a href=\"/code/2013/testing-frontend-javascript-code-using-mocha-chai-and-sinon/\">in a previous episode</a>:</p>\n<pre><code>// cow.js\n(function(exports) {\n  \"use strict\";\n\n  function Cow(name) {\n    this.name = name || \"Anon cow\";\n  }\n  exports.Cow = Cow;\n\n  Cow.prototype = {\n    greets: function(target) {\n      if (!target)\n        throw new Error(\"missing target\");\n      return this.name + \" greets \" + target;\n    }\n  };\n})(this);\n</code></pre>\n\n                        <p>And its test suite, powered by Mocha and <a href=\"http://chaijs.com/\">Chai</a>:</p>\n<pre><code>var expect = chai.expect;\n\ndescribe(\"Cow\", function() {\n  describe(\"constructor\", function() {\n    it(\"should have a default name\",function() {\n      var cow = new Cow();\n      expect(cow.name).to.equal(\"Anon cow\");\n    });\n\n    it(\"should set cow's name if provided\", function() {\n      var cow = new Cow(\"Kate\");\n      expect(cow.name).to.equal(\"Kate\");\n    });\n  });\n\n  describe(\"#greets\", function() {\n    it(\"should greet passed target\", function() {\n      var greetings = (new Cow(\"Kate\")).greets(\"Baby\");\n      expect(greetings).to.equal(\"Kate greets Baby\");\n    });\n  });\n});\n</code></pre>\n\n                        <p>Let's create the HTML test file for it, featuring Blanket and its adapter\n                            for Mocha:</p>\n<pre><code>&lt;!DOCTYPE html&gt;\n&lt;html&gt;\n&lt;head&gt;\n  &lt;meta charset=\"utf-8\"&gt;\n  &lt;title&gt;Test&lt;/title&gt;\n  &lt;link rel=\"stylesheet\" media=\"all\" href=\"vendor/mocha.css\"&gt;\n&lt;/head&gt;\n&lt;body&gt;\n  &lt;div id=\"mocha\"&gt;&lt;/div&gt;\n  &lt;div id=\"messages\"&gt;&lt;/div&gt;\n  &lt;div id=\"fixtures\"&gt;&lt;/div&gt;\n  &lt;script src=\"vendor/mocha.js\"&gt;&lt;/script&gt;\n  &lt;script src=\"vendor/chai.js\"&gt;&lt;/script&gt;\n  &lt;script src=\"vendor/blanket.js\"\n          data-cover-adapter=\"vendor/mocha-blanket.js\"&gt;&lt;/script&gt;\n  &lt;script&gt;mocha.setup('bdd');&lt;/script&gt;\n  &lt;script src=\"cow.js\" data-cover&gt;&lt;/script&gt;\n  &lt;script src=\"cow_test.js\"&gt;&lt;/script&gt;\n  &lt;script&gt;mocha.run();&lt;/script&gt;\n&lt;/body&gt;\n&lt;/html&gt;\n</code></pre>\n\n                        <p><strong>Notes</strong>:</p>\n                        <ul>\n                            <li>Notice the <code>data-cover</code> attribute we added to the script tag\n                                loading the source of our library;</li>\n                            <li>The HTML test file <em>must</em> be served over HTTP for the adapter to\n                                be loaded.</li>\n                        </ul>\n                        <p>Running the tests now gives us something like this:</p>\n                        <p>\n                            <img alt=\"screenshot\" src=\"/static/code/2013/blanket-coverage.png\">\n                        </p>\n                        <p>As you can see, the report at the bottom highlights that we haven't actually\n                            tested the case where an error is raised in case a target name is missing.\n                            We've been informed of that, nothing more, nothing less. We simply know\n                            we're missing a test here. Isn't this cool? I think so!</p>\n                        <p>Just remember that code coverage will only <a href=\"http://codebetter.com/karlseguin/2008/12/09/code-coverage-use-it-wisely/\">bring you numbers</a> and\n                            raw information, not actual proofs that the whole of your <em>code logic</em> has\n                            been actually covered. If you ask me, the best inputs you can get about\n                            your code logic and implementation ever are the ones issued out of <a href=\"http://www.extremeprogramming.org/rules/pair.html\">pair programming</a>\nsessions\n                            and <a href=\"http://alexgaynor.net/2013/sep/26/effective-code-review/\">code reviews</a> —\n                            but that's another story.</p>\n                        <p><strong>So is code coverage silver bullet? No. Is it useful? Definitely. Happy testing!</strong>\n                        </p>\n                    </section></div></div>",
  "text_content": "So finally you're testing your frontend JavaScript code ? Great! The more you\nwrite tests, the more confident you are with your code… but how much precisely?\nThat's where code coverage might\nhelp. The idea behind code coverage is to record which parts of your code (functions, statements, conditionals and so on) have been executed by your test suite, to compute metrics out of these data and usually to provide tools for navigating and inspecting them. Not a lot of frontend developers I know actually test their frontend code, and I can barely imagine how many of them have ever setup code coverage… Mostly because there are not many frontend-oriented tools in this area I guess. Actually I've only found one which provides an adapter for Mocha and actually works… Drinking game for web devs: (1) Think of a noun (2) Google \"<noun>.js\" (3) If a library with that name exists - drink — Shay Friedman (@ironshay) August 22, 2013 Blanket.js is an easy to install, easy to configure,\nand easy to use JavaScript code coverage library that works both in-browser and\nwith nodejs. Its use is dead easy, adding Blanket support to your Mocha test suite is just matter of adding this simple line to your HTML test file: <script src=\"vendor/blanket.js\" data-cover-adapter=\"vendor/mocha-blanket.js\"></script> Source files: blanket.js , mocha-blanket.js As an example, let's reuse the silly Cow example we used in a previous episode : // cow.js\n(function(exports) { \"use strict\"; function Cow(name) { this.name = name || \"Anon cow\"; } exports.Cow = Cow; Cow.prototype = { greets: function(target) { if (!target) throw new Error(\"missing target\"); return this.name + \" greets \" + target; } };\n})(this); And its test suite, powered by Mocha and Chai : var expect = chai.expect; describe(\"Cow\", function() { describe(\"constructor\", function() { it(\"should have a default name\",function() { var cow = new Cow(); expect(cow.name).to.equal(\"Anon cow\"); }); it(\"should set cow's name if provided\", function() { var cow = new Cow(\"Kate\"); expect(cow.name).to.equal(\"Kate\"); }); }); describe(\"#greets\", function() { it(\"should greet passed target\", function() { var greetings = (new Cow(\"Kate\")).greets(\"Baby\"); expect(greetings).to.equal(\"Kate greets Baby\"); }); });\n}); Let's create the HTML test file for it, featuring Blanket and its adapter for Mocha: <!DOCTYPE html>\n<html>\n<head> <meta charset=\"utf-8\"> <title>Test</title> <link rel=\"stylesheet\" media=\"all\" href=\"vendor/mocha.css\">\n</head>\n<body> <div id=\"mocha\"></div> <div id=\"messages\"></div> <div id=\"fixtures\"></div> <script src=\"vendor/mocha.js\"></script> <script src=\"vendor/chai.js\"></script> <script src=\"vendor/blanket.js\" data-cover-adapter=\"vendor/mocha-blanket.js\"></script> <script>mocha.setup('bdd');</script> <script src=\"cow.js\" data-cover></script> <script src=\"cow_test.js\"></script> <script>mocha.run();</script>\n</body>\n</html> Notes : Notice the data-cover attribute we added to the script tag loading the source of our library; The HTML test file must be served over HTTP for the adapter to be loaded. Running the tests now gives us something like this: As you can see, the report at the bottom highlights that we haven't actually tested the case where an error is raised in case a target name is missing. We've been informed of that, nothing more, nothing less. We simply know we're missing a test here. Isn't this cool? I think so! Just remember that code coverage will only bring you numbers and raw information, not actual proofs that the whole of your code logic has been actually covered. If you ask me, the best inputs you can get about your code logic and implementation ever are the ones issued out of pair programming sessions and code reviews — but that's another story. So is code coverage silver bullet? No. Is it useful? Definitely. Happy testing!",
  "length": 3821,
  "excerpt": "Nicolas Perriault's homepage.",
  "site_name": null,
  "published_time": null
}
